crates/bevy_ecs/src/bundle.rs

```diff
@@ -905,16 +928,28 @@ impl<'w> BundleSpawner<'w> {
 
         // SAFETY: We have no outstanding mutable references to world as they were dropped
         let mut deferred_world = unsafe { self.world.into_deferred() };
-        if archetype.has_on_add() {
-            // SAFETY: All components in the bundle are guaranteed to exist in the World
-            // as they must be initialized before creating the BundleInfo.
-            unsafe { deferred_world.trigger_on_add(entity, bundle_info.iter_components()) };
-        }
-        if archetype.has_on_insert() {
-            // SAFETY: All components in the bundle are guaranteed to exist in the World
-            // as they must be initialized before creating the BundleInfo.
-            unsafe { deferred_world.trigger_on_insert(entity, bundle_info.iter_components()) };
-        }
+        // SAFETY: `DeferredWorld` cannot provide mutable access to `Archetypes`.
+        let archetype = self.archetype.as_ref();
+        // SAFETY: All components in the bundle are guaranteed to exist in the World
+        // as they must be initialized before creating the BundleInfo.
+        unsafe {
+            deferred_world.trigger_on_add(archetype, entity, bundle_info.iter_components());
+            if archetype.has_add_observer() {
+                deferred_world.trigger_observers(
+                    ON_ADD,
+                    Some(entity),
+                    bundle_info.iter_components(),
+                );
+            }
+            deferred_world.trigger_on_insert(archetype, entity, bundle_info.iter_components());
+            if archetype.has_insert_observer() {
+                deferred_world.trigger_observers(
+                    ON_INSERT,
+                    Some(entity),
+                    bundle_info.iter_components(),
+                );
+            }
+        };
```

Comment:
On experimentation I remembered why I didn't factor this out in the first place. You can't just pass the bundle_info down since sometimes you are using the add_bundle instead and you can't just pass the iterator down because you need to consume it twice to trigger hooks and then observers. It's a bit annoying.

Reply:
I've removed the duplicate code, minus the one instance that is slightly different.